addons/api/addon/services/indexed-db.js

```diff
@@ -0,0 +1,114 @@
+import Service, { inject as service } from '@ember/service';
+import Dexie from 'dexie';
+import { typeOf } from '@ember/utils';
+
+// List of resources we support to be cached and searched in
+// indexedDb. Any field that can be searched upon should be added to
+// this index. Increment the version number of the database whenever
+// indexes change.
+export const modelIndexes = {
+  tokens: '&id, token',
+  targets: '&id, attributes.type, attributes.name, attributes.scope.scope_id',
+  sessions: '&id, attributes.type, attributes.scope.scope_id',
+};
+
+export const formatDbName = (userId, clusterUrl) =>
+  `boundary-${userId}-${clusterUrl}`;
+
+/**
+ * Recursively copy over the object and convert any booleans to "true" or "false" so they can be indexed
+ * @param result
+ * @param object
+ * @returns {*}
+ */
+const convertBooleansInObject = (result, object) => {
+  Object.entries(object).forEach(([key, value]) => {
+    if (typeOf(value) === 'object') {
+      result[key] = convertBooleansInObject({}, value);
+    } else if (typeOf(value) === 'boolean') {
+      result[key] = value ? 'true' : 'false';
+    } else {
+      result[key] = value;
+    }
+  });
+
+  return result;
+};
+
+/**
+ * Recursively copy over the object and convert any string "true" or "false" back to booleans
+ * @param result
+ * @param object
+ * @returns {*}
+ */
+const unconvertBooleansInObject = (result, object) => {
+  Object.entries(object).forEach(([key, value]) => {
+    if (typeOf(value) === 'object') {
+      result[key] = unconvertBooleansInObject({}, value);
+    } else if (value === 'true' || value === 'false') {
+      result[key] = value === 'true' ? true : false;
+    } else {
+      result[key] = value;
+    }
+  });
+
+  return result;
+};
+
+/**
+ * Service to encapsulate the IndexedDB implementation. To use this service, call
+ * `setup` from the application root and
+ */
+export default class IndexedDbService extends Service {
+  // =services
+  @service session;
+
+  // =attributes
+  #db;
+  get db() {
+    return this.#db;
+  }
+
+  setup(dbName) {
+    // Don't run setup again if we already have one or if we didn't get a name
+    if (this.#db || !dbName) {
+      return;
+    }
+
+    this.#db = new Dexie(dbName);
+    this.#db.version(1).stores(modelIndexes);
+  }
+
+  /**
+   * Cleanup the data by converting all booleans to "false" or "true".
+   * IndexedDB doesn't support indexes on boolean values so we have to
+   * convert them to be able to support using boolean values.
+   *
+   * Data is assumed to be in POJOs and that our normalized data are not
+   * ember objects, otherwise they should also be converted.
+   *
+   * Undo the operation by setting `convertBooleans` to false.
+   *
+   * @param {Object} data
+   * @param {Boolean} convertBooleans
+   * @return {{id, type, attributes: {}, relationships: {}}}
+   */
+  normalizeData(data, convertBooleans) {
```

Comment:
Unfortunately indexedDb doesn't support booleans as part of their index so I need this hack to convert all booleans to strings so we can index on them.

I'm not exactly sure how big of a performance hit this is having to convert back and forth every time and if it turns out to be significant, we can either not index on boolean fields or create a custom mapper for each resource where we can specify which fields are booleans and save these values as extra fields to be stored as 0 or 1 instead which we can use to index on instead. 

e.g. This is what the data would look like in that case in indexed db. The annoying part is having to now maintain the mapping of models and changing this anytime we change our models that involve booleans, which is why I didn't start off with this. 
```
id: '123',
attributes: {}, // This is the normal data which would include the actual fields
boolean1: 1,
boolean2: 0
```